brave-kafka/src/main/java/smartthings/brave/kafka/consumers/TracingConsumerInterceptor.java

```diff
@@ -0,0 +1,129 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package smartthings.brave.kafka.consumers;
+
+import brave.Tracer;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+import com.google.protobuf.InvalidProtocolBufferException;
+import org.apache.kafka.clients.consumer.ConsumerInterceptor;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.config.ConfigException;
+import smartthings.brave.kafka.EnvelopeProtos;
+import zipkin.Constants;
+import zipkin.Endpoint;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TracingConsumerInterceptor<K> implements ConsumerInterceptor<K, byte[]> {
+
+  private Tracer tracer;
+  private SpanNameProvider<K> nameProvider;
+  private final Endpoint kafkaEndpoint;
+  private final static Logger logger = Logger.getLogger(TracingConsumerInterceptor.class.getName());
+
+  // TODO: how to inject dependencies without using config
+
+  public TracingConsumerInterceptor() {
+    kafkaEndpoint = Endpoint.builder().serviceName("Kafka").build();
+  }
+
+  private final TraceContext.Extractor<EnvelopeProtos.Envelope> extractor = new TraceContext.Extractor<EnvelopeProtos.Envelope>() {
+    @Override
+    public TraceContextOrSamplingFlags extract(EnvelopeProtos.Envelope envelope) {
+      TraceContext.Builder builder = TraceContext.newBuilder()
+        .traceIdHigh(envelope.getTraceIdHigh())
+        .traceId(envelope.getTraceId())
+        .spanId(envelope.getSpanId())
+        .sampled(envelope.getSampled())
+        .shared(envelope.getShared());
+      if (envelope.getParentId() != 0) {
+        builder.parentId(envelope.getParentId());
+      }
+      return TraceContextOrSamplingFlags.create(builder);
+    }
+  };
+
+
+  @Override
+  public ConsumerRecords<K, byte[]> onConsume(ConsumerRecords<K, byte[]> records) {
+    Map<TopicPartition, List<ConsumerRecord<K, byte[]>>> unenvelopedRecords = new HashMap<>(records.count());
+    records.forEach(record -> {
+      try {
+        EnvelopeProtos.Envelope envelope = EnvelopeProtos.Envelope.parseFrom(record.value());
+        TraceContextOrSamplingFlags traceContextOrSamplingFlags = extractor.extract(envelope);
+        TraceContext ctx = traceContextOrSamplingFlags.context();
+        (ctx != null ? tracer.joinSpan(ctx) : tracer.newTrace(traceContextOrSamplingFlags.samplingFlags()))
+          .name(nameProvider.spanName(record))
+          .annotate(Constants.SERVER_RECV)
+          .remoteEndpoint(kafkaEndpoint)
+          .tag("Partition", record.partition() + "")
+          .tag("Offset", record.offset() + "")
```

Comment:
Should use lowercase dot delimited key name.  Maybe "kafka.offset".  Also unless these are useful for searching we might not way to even store these.  Tags do incur a cost so we just want to make sure they are going to be useful.